unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,432 @@
+use cid::Cid;
+
+use crate::pb::{FlatUnixFs, PBLink, UnixFs, UnixFsType};
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+use std::num::NonZeroUsize;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder
+pub struct FileAdder {
+    chunker: Chunker,
+    block_buffer: Vec<u8>,
+    // the index in the outer vec is the height (0 == leaf, 1 == first links to leafs, 2 == links
```

Comment:
is the maximum outer index `2`, i.e. `length == 3`? in that case we might want to consider something like `smallvec`

Reply:
Rewrote this and moved it into the `BalancedCollector`. the current way is just using this as a flat vec, which holds compressable items. it's compacted from the end to the start.